// cylon/dist/robot.js
/*
 * robot
 * cylonjs.com
 *
 * Copyright (c) 2013 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/


(function() {
  'use strict';
  var Async, namespace,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  require('./cylon');

  require('./basestar');

  require("./connection");

  require("./adaptor");

  require("./device");

  require("./driver");

  require('./digital-pin');

  namespace = require('node-namespace');

  Async = require("async");

  namespace('Cylon', function() {
    return this.Robot = (function() {
      var klass;

      klass = Robot;

      function Robot(opts) {
        var func, n, reserved;
        if (opts == null) {
          opts = {};
        }
        this.toString = __bind(this.toString, this);
        this.registerDriver = __bind(this.registerDriver, this);
        this.requireDriver = __bind(this.requireDriver, this);
        this.registerAdaptor = __bind(this.registerAdaptor, this);
        this.requireAdaptor = __bind(this.requireAdaptor, this);
        this.stop = __bind(this.stop, this);
        this.startDevices = __bind(this.startDevices, this);
        this.startConnections = __bind(this.startConnections, this);
        this.start = __bind(this.start, this);
        this.initDevices = __bind(this.initDevices, this);
        this.initConnections = __bind(this.initConnections, this);
        this.robot = this;
        //this.name = opts.name || this.constructor.randomName();
        this.name = "test"
        this.master = opts.master;
        this.connections = {};
        this.devices = {};
        this.adaptors = {};
        this.drivers = {};
        this.commands = [];
        //this.registerAdaptor("./test/loopback", "loopback");
        //this.registerDriver("./test/ping", "ping");
        this.initConnections(opts.connection || opts.connections);
        Logger.info("opts.device", opts.device);
        Logger.info("opts.devices", opts.devices);
        this.initDevices(opts.device || opts.devices);
        this.work = opts.work || function() {
          return Logger.info("No work yet");
        };
        for (n in opts) {
          func = opts[n];
          reserved = ['connection', 'connections', 'device', 'devices', 'work'];
          if (__indexOf.call(reserved, n) < 0) {
            this.robot[n] = func;
          }
        }
      }

      Robot.randomName = function() {
        return "Robot " + (Math.floor(Math.random() * 100000));
      };

      Robot.prototype.data = function() {
        var connection, device, n;
        return {
          name: this.name,
          connections: (function() {
            var _ref, _results;
            _ref = this.connections;
            _results = [];
            for (n in _ref) {
              connection = _ref[n];
              _results.push(connection.data());
            }
            return _results;
          }).call(this),
          devices: (function() {
            var _ref, _results;
            _ref = this.devices;
            _results = [];
            for (n in _ref) {
              device = _ref[n];
              _results.push(device.data());
            }
            return _results;
          }).call(this),
          commands: this.commands
        };
      };

      Robot.prototype.initConnections = function(connections) {
        var connection, _i, _len;
	_len = connections.length
        Logger.info("Initializing connections...");
        if (connections == null) {
          return;
        }
        Logger.info("1")
        //connections = [].concat(connections);
	var c = []
	c.push(connections)
        Logger.info("2")
        for (_i = 0, _len = c.length; _i < _len; _i++) {
          Logger.info(i)
          connection = c[0];
          Logger.info("Initializing connection '" + connection.name + "'...");
          connection['robot'] = this;
          this.connections[connection.name] = new Cylon.Connection(connection);
        }
        return this.connections;
      };

      Robot.prototype.initDevices = function(devices) {
        var device, _i, _len, _results;
        Logger.info("Initializing devices...");
        if (devices == null) {
          return;
        }
        //devices = [].concat(devices);
	var d = []
	d.push(devices)
	Logger.info("d", d);
	Logger.info("d[0]", d[0]);
        _results = [];
        for (_i = 0, _len = d.length; _i < _len; _i++) {
	  Logger.info("for d", d);
	  Logger.info("d[_i]", d[0]);
          Logger.info("initDevies 1");
          //device = d[_i];
          Logger.info("initDevices 2", d[0]);
          //Logger.info("Initializing device '" + device.name + "'...");
          Logger.info("Initializing device '" + d[0].name + "'...");
          Logger.info("initDevices 3");
          //device['robot'] = this;
          d[0]['robot'] = this;
          Logger.info("initDevices 4");
          //_results.push(this.devices[device.name] = new Cylon.Device(device));
          _results.push(this.devices[d[0].name] = new Cylon.Device(d[0]));
          Logger.info("initDevices 5");
        }
        return _results;
      };

      Robot.prototype.start = function() {
        var this1 = this;
        Logger.info("start 1");
        return this.startConnections(function() {
        Logger.info("start 2");
          return this1.robot.startDevices(function() {
        Logger.info("start 3");
            return this1.robot.work.call(this1.robot, this1.robot);
        Logger.info("start 4");
          });
        });
      };

      Robot.prototype.startConnections = function(callback) {
        var connection, n, starters, _ref;
        Logger.info("Starting connections...");
        starters = {};
        _ref = this.connections;
        for (n in _ref) {
          connection = _ref[n];
          starters[n] = connection.connect;
        }
        return Async.parallel(starters, callback);
      };

      Robot.prototype.startDevices = function(callback) {
        var device, n, starters, ref;
        Logger.info("Starting devices...");
        starters = {};
        Logger.info("startDevices 1");
        ref = this.devices;
        Logger.info("startDevices 2");
        for (n in ref) {
	  Logger.info("n", n);
          Logger.info("startDevices 3");
	  Logger.info("_ref[n]", ref[n]);
	  Logger.info("_ref[n]", ref[n]);
          device = ref[n];
          Logger.info("startDevices 4");
          this.robot[n] = device;
          Logger.info("startDevices 5");
          starters[n] = device.start;
          Logger.info("startDevices 6");
        }
        Logger.info("startDevices 4");
        return Async.parallel(starters, callback);
      };

      Robot.prototype.stop = function() {
        var connection, device, n, _ref, _ref1, _results;
        _ref = this.devices;
        for (n in _ref) {
          device = _ref[n];
          device.stop();
        }
        _ref1 = this.connections;
        _results = [];
        for (n in _ref1) {
          connection = _ref1[n];
          _results.push(connection.disconnect());
        }
        return _results;
      };

      Robot.prototype.initAdaptor = function(adaptorName, connection, opts) {
        if (opts == null) {
          opts = {};
        }
        return this.robot.requireAdaptor(adaptorName).adaptor({
          name: adaptorName,
          connection: connection,
          extraParams: opts
        });
      };

      Robot.prototype.requireAdaptor = function(adaptorName) {
        if (this.robot.adaptors[adaptorName] == null) {
          this.robot.registerAdaptor("cylon-" + adaptorName, adaptorName);
          this.robot.adaptors[adaptorName].register(this);
        }
        return this.robot.adaptors[adaptorName];
      };

      Robot.prototype.registerAdaptor = function(moduleName, adaptorName) {
        if (this.adaptors[adaptorName] == null) {
          return this.adaptors[adaptorName] = require(moduleName);
        }
      };

      Robot.prototype.initDriver = function(driverName, device, opts) {
        if (opts == null) {
          opts = {};
        }
        return this.robot.requireDriver(driverName).driver({
          name: driverName,
          device: device,
          extraParams: opts
        });
      };

      Robot.prototype.requireDriver = function(driverName) {
        if (this.robot.drivers[driverName] == null) {
          this.robot.registerDriver("cylon-" + driverName, driverName);
          this.robot.drivers[driverName].register(this);
        }
        return this.robot.drivers[driverName];
      };

      Robot.prototype.registerDriver = function(moduleName, driverName) {
        if (this.drivers[driverName] == null) {
          return this.drivers[driverName] = require(moduleName);
        }
      };

      Robot.prototype.toString = function() {
        return "[Robot name='" + this.name + "']";
      };

      return Robot;

    })();
  });

  module.exports = Cylon.Robot;

}).call(this);
